tree/treeplayer/src/TMPWorkerTree.cxx

```diff
@@ -242,20 +239,63 @@ void TMPWorkerTreeSel::Process(unsigned int code, MPCodeBufPair& msg)
 {
    //evaluate the index of the file to process in fFileNames
    //(we actually don't need the parameter if code == kProcTree)
-   unsigned fileN = 0;
-   unsigned nProcessed = 0;
-   TTree *tree = 0;
+
    Long64_t start = 0;
    Long64_t finish = 0;
+   TEntryList *enl = 0;
+   std::string errmsg;
+   if (LoadTree(code, msg, start, finish, &enl, errmsg) != 0) {
+      SendError(errmsg);
+      return;
+   }
+
+   if (fCallBegin) {
+      fSelector.SlaveBegin(nullptr);
+      fCallBegin = false;
+   }
+
+   fSelector.Init(fTree);
+   fSelector.Notify();
+   for (Long64_t entry = start; entry < finish; ++entry) {
+      Long64_t e = (enl) ? enl->GetEntry(entry) : entry;
+      fSelector.Process(e);
+   }
+
+   // update the number of processed entries
+   fProcessedEntries += finish - start;
+
+   MPSend(GetSocket(), MPCode::kIdling);
+
+   return;
+}
+
+/// Load the requierd tree and evaluate the processing range
+
+Int_t TMPWorkerTree::LoadTree(unsigned int code, MPCodeBufPair &msg, Long64_t &start, Long64_t &finish,
+                              TEntryList **enl, std::string &errmsg)
+{
+   // evaluate the index of the file to process in fFileNames
+   //(we actually don't need the parameter if code == kProcTree)
+
+   start = 0;
+   finish = 0;
+   errmsg = "";
+
+   unsigned fileN = 0;
+   unsigned nProcessed = 0;
    bool setupcache = true;
 
+   std::string mgroot = "[S" + std::to_string(GetNWorker()) + "]: ";
+
+   TTree *tree = 0;
    if (code ==  MPCode::kProcTree) {
 
+      mgroot += "MPCode::kProcTree: ";
+
       // The tree must be defined at this level
       if(fTree == nullptr) {
-         std::cout << "tree undefined!\n" ;
```

Comment:
Hi Philippe,
On the side: all std::cout and alike are already remove from the MP code (the line the you indicate is before the PR). But checking for this I have found plenty of places in treereader where std::cout is used. For example:
$ grep -n "cout" tree/treeplayer/inc/*h
tree/treeplayer/inc/TBranchProxy.h:314:         std::cout << "fWhere " << fWhere << std::endl;
tree/treeplayer/inc/TBranchProxy.h:315:         if (fWhere) std::cout << "value? " << *(unsigned char*)GetStart() << std::endl;
tree/treeplayer/inc/TBranchProxy.h:372:         std::cout << "fWhere " << fWhere << std::endl;
tree/treeplayer/inc/TBranchProxy.h:375:               std::cout << "location " << *(TClonesArray**)fWhere << std::endl;
tree/treeplayer/inc/TBranchProxy.h:377:               std::cout << "location " << fWhere << std::endl;
tree/treeplayer/inc/TBranchProxy.h:414:         std::cout << "fWhere " << fWhere << std::endl;
tree/treeplayer/inc/TBranchProxy.h:417:               std::cout << "location " << *(TClonesArray**)fWhere << std::endl;
tree/treeplayer/inc/TBranchProxy.h:419:               std::cout << "location " << fWhere << std::endl;
tree/treeplayer/inc/TBranchProxy.h:455:         std::cout << "fWhere " << fWhere << std::endl;
tree/treeplayer/inc/TBranchProxy.h:456:         if (fWhere) std::cout << "value? " << *(T*)GetStart() << std::endl;
tree/treeplayer/inc/TBranchProxy.h:529:         std::cout << "fWhere " << GetWhere() << std::endl;
tree/treeplayer/inc/TBranchProxy.h:530:         if (GetWhere()) std::cout << "value? " << *(type_t*)GetWhere() << std::endl;
tree/treeplayer/inc/TBranchProxy.h:555:         std::cout << "fWhere " << fWhere << std::endl;
tree/treeplayer/inc/TBranchProxy.h:556:         if (fWhere) std::cout << "value? " << *(T*)GetStart() << std::endl;
tree/treeplayer/inc/TBranchProxy.h:607:         std::cout << "fWhere " << fWhere << std::endl;
tree/treeplayer/inc/TBranchProxy.h:608:         if (fWhere) std::cout << "value? " << *(T*)GetStart() << std::endl;
tree/treeplayer/inc/TBranchProxy.h:660:         std::cout << "fWhere " << fWhere << std::endl;
tree/treeplayer/inc/TBranchProxy.h:661:         if (fWhere) std::cout << "value? " << *(type_t*)GetStart() << std::endl;
tree/treeplayer/inc/TBranchProxy.h:697:         std::cout << "fWhere " << fWhere << std::endl;
tree/treeplayer/inc/TBranchProxy.h:698:         if (fWhere) std::cout << "value? " << *(type_t*)GetStart() << std::endl;
tree/treeplayer/inc/TBranchProxyTemplate.h:52:         std::cout << "fWhere " << obj.GetWhere() << std::endl;
tree/treeplayer/inc/TBranchProxyTemplate.h:53:         if (obj.GetWhere()) std::cout << "address? " << (T*)obj.GetWhere() << std::endl;
tree/treeplayer/inc/TBranchProxyTemplate.h:78:         std::cout << "obj.GetWhere() " << obj.GetWhere() << std::endl;
tree/treeplayer/inc/TBranchProxyTemplate.h:79:         //if (obj.GetWhere()) std::cout << "value? " << *(T*)obj.GetWhere() << std::endl;
tree/treeplayer/inc/TBranchProxyTemplate.h:120:         std::cout << "obj.GetWhere() " << obj.GetWhere() << std::endl;
tree/treeplayer/inc/TBranchProxyTemplate.h:121:         //if (obj.GetWhere()) std::cout << "value? " << *(T*)obj.GetWhere() << std::endl;
 
I do not know if we have to start a campaign to remove all those ...